tensorflow_addons/layers/decorators_test.py

```diff
@@ -0,0 +1,176 @@
+    def fn_use_data_dep(x):
+      return m(x)
+
+    @decorators.recompute_grad(tupleize_grads=True)
+    def fn_tupleize(x):
+      return m(x)
+
+    @decorators.recompute_grad(use_data_dep=True, tupleize_grads=True)
+    def fn_both(x):
+      return m(x)
+
+    x = tf.random.uniform((3, 1, 3))
+
+    names_and_fns = [
+        ("recompute", fn_recompute),
+        ("regular", fn),
+        ("use_data_dep", fn_use_data_dep),
+        ("tupleize", fn_tupleize),
+        ("tuple_and_data_dep", fn_both),
+    ]
+    outputs_and_vars = []
+    all_grads = []
+    for name, wrapped_fn in names_and_fns:
+      with tf.GradientTape() as tape:
+        out = tf.math.reduce_sum(wrapped_fn(x))
+      outputs_and_vars.append((out, m.trainable_variables))
+      all_grads.append(tape.gradient(out, m.trainable_variables))
+
+    outputs = list(zip(*outputs_and_vars))[0]
+
+    # All outputs are the same
+    current = outputs[0]
+    for out in outputs[1:]:
+      self.assertAllClose(current, out)
+      current = out
+
+    # All gradients are the same
+    for grads in zip(all_grads):
+      current = grads[0]
+      for g in grads[1:]:
+        self.assertAllClose(current, g)
+        current = g
+
+  def testDoubleCall(self):
+
+    dense = tf.keras.layers.Dense(2)
+
+    @decorators.recompute_grad
+    def layer_with_recompute(inputs):
+      return dense(inputs)
+
+    with tf.GradientTape() as tape:
+      inputs = tf.ones((2, 4), tf.float32)
+      out1 = layer_with_recompute(inputs)
+      out2 = layer_with_recompute(inputs) + out1
+      out = tf.math.reduce_sum(out2)
+
+    tvars = dense.trainable_variables
+    assert len(tvars) == 2
+    grads = tape.gradient(out, tvars)
+    for grad in grads:
+      self.assertTrue(grad is not None)
+
+  def testWithIsRecomputeKwarg(self):
+
+    kwarg_values = []
+
+    dense = tf.keras.layers.Dense(2)
+    batch_normalization = tf.keras.layers.BatchNormalization()
+
+    @decorators.recompute_grad
+    def layer_with_recompute(inputs, is_recomputing=False):
+      kwarg_values.append(is_recomputing)
+
+      out = dense(inputs)
+      out = batch_normalization(out, training=True)
+      # TODO: What to do with this? How do we keep the BatchNormalization
+      # layer from updating the statistics but still using only batch statistics?
+      #if is_recomputing:
+      #  # Ensure that the updates are not duplicated by popping off the latest
+      #  # 2 additions.
+      #  update_ops = ops.get_collection_ref(ops.GraphKeys.UPDATE_OPS)
+      #  update_ops.pop()
+      #  update_ops.pop()
+      return out
+
+    with tf.GradientTape() as tape:
+      x = tf.ones((2, 4), tf.float32)
+      y = layer_with_recompute(x)
+      loss = tf.math.reduce_sum(y)
+    tvars = dense.trainable_variables + \
+            batch_normalization.trainable_variables
+    tape.gradient(loss, tvars)
+
+    # TODO: What to do with this?
+    #update_ops = ops.get_collection(ops.GraphKeys.UPDATE_OPS)
```

Comment:
Hi, Niels, I'm not sure what do you mean here. If we want to collect update_ops, perhaps we could build [tf.keras.Sequential](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/Sequential) or create a subclass of [tf.keras.Model](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/Model). And then collect all update_ops by `[layer.updates for layer in model.layers]`, does it work for you?

Reply:
You might want to read: https://www.tensorflow.org/alpha/guide/keras/functional

Reply:
I mean what this check is supposed to do the following: Because of the recompute_grad decorator the function is evaluated multiple times. But we want to only update the batch normalization statistics on the first run, because otherwise we get wrong statistics.
So far in graph mode the batch normalization updates were added to the UPDATE_OPS collection. Therefore the code above could remove the duplicate update ops before they were carried out.

But now in eager mode the updates are carried out in place. So we can't take them back anymore.
Therefore we need a way to avoid that the update happens in the first place in eager mode.
I don't know how to do that without modifying the BatchNormalization layer so this still is a TODO.

Reply:
Got it, thank you for the clarification, Niels. 

I'll be out of town in the next two weeks, @WindQAQ @kyleabeauchamp Tzu-Wei, Kyle, could you take a look and help Niels out when time allows? Thank you :-)

Reply:
@pavithrasv @qlzh727 @mihaimaruseac Hi, Pavithra, Qianli, Mihai, do you have any good idea to control the updating behavior of layer? Thank you very much, cc @fchollet